cvat/apps/consensus/new_intersect_merge.py

```diff
@@ -0,0 +1,1107 @@
+    ):
+        if label_matcher is None:
+            label_matcher = self.label_matcher
+        if dist_thresh is None:
+            dist_thresh = self.pairwise_dist
+        if a_objs is None:
+            a_objs = self._get_ann_type(t, item_a)
+        if b_objs is None:
+            b_objs = self._get_ann_type(t, item_b)
+
+        if self.return_distances:
+            distance, distances = self._make_memoizing_distance(distance)
+
+        if not a_objs and not b_objs:
+            distances = {}
+            returned_values = [], [], [], []
+        else:
+            extra_args = {}
+            if label_matcher:
+                extra_args["label_matcher"] = label_matcher
+
+            returned_values = match_segments(
+                a_objs,
+                b_objs,
+                distance=distance,
+                dist_thresh=dist_thresh if dist_thresh is not None else self.iou_threshold,
+                **extra_args,
+            )
+
+        if self.return_distances:
+            returned_values = returned_values + (distances,)
+
+        return returned_values
+
+    def match_annotations_two_sources(self, a, b):
+        pass
+
+    def match_annotations(self, sources):
+        distance = self.distance
+        pairwise_dist = self.pairwise_dist
+        cluster_dist = self.cluster_dist
+
+        if cluster_dist < 0:
+            cluster_dist = pairwise_dist
+
+        id_segm = {id(a): (a, id(s)) for s in sources for a in s}
+
+        def _is_close_enough(cluster, extra_id):
+            # check if whole cluster IoU will not be broken
+            # when this segment is added
+            b = id_segm[extra_id][0]
+            for a_id in cluster:
+                a = id_segm[a_id][0]
+                if distance(a, b) < cluster_dist:
+                    return False
+            return True
+
+        def _has_same_source(cluster, extra_id):
+            b = id_segm[extra_id][1]
+            for a_id in cluster:
+                a = id_segm[a_id][1]
+                if a == b:
+                    return True
+            return False
+
+        # match segments in sources, pairwise
+        adjacent = {i: [] for i in id_segm}  # id(sgm) -> [id(adj_sgm1), ...]
+        for a_idx, src_a in enumerate(sources):
+            # matches further sources of same frame for matching annotations
+            for src_b in sources[a_idx + 1 :]:
+                # an annotation can be adjacent to multiple annotations
+                if self.return_distances:
+                    matches, _, _, _, _ = self.match_annotations_two_sources(
+                        src_a,
+                        src_b,
+                    )
+                else:
+                    matches, _, _, _ = self.match_annotations_two_sources(
+                        src_a,
+                        src_b,
+                    )
+                for a, b in matches:
+                    adjacent[id(a)].append(id(b))
```

Comment:
What does this code do?

Reply:
This finds all matching annotations exist in the remaining sources for an annotation `a`. Based on this only the cluster formation and subsequently merging happens.

Reply:
If it's a class method, why it returns these 2 kinds of output? I seems like only one of them is used, and so I suggest that the function is refactored to return only 1 kind of resulting value.

Reply:
Two kinds of output?

Reply:
Are you talking about match annotations two sources?

Reply:
Yes, about this `if` with 2 possible output types with 4 or 5 values:
```
if self.return_distances:
                    matches, _, _, _, _ = self.match_annotations_two_sources(
                        src_a,
                        src_b,
                    )
else:
                    matches, _, _, _ = self.match_annotations_two_sources(
                        src_a,
                        src_b,
                    )
```

Reply:
In skeleton matcher the self.return distance is true and there we use the 5th return value

Reply:
Instead we can make it return only two values matches and distance. The distance will be empty dict when self.return distance is false

Reply:
So, does any of the classes in the matcher or merger hierarchies uses the returned distances? If no, I don't see a point why this class method returns different value sets.

Reply:
The one in skeleton matcher

Reply:
I can't find the specific code that uses the 5th value from a `self.match_annotations_two_sources()` call. In skeleton merger the distances are just remembered in a class field.

Reply:
https://github.com/cvat-ai/cvat/blob/11ef2707eb3681bc0b821816ae74bb6af7c2d7dc/cvat/apps/consensus/new_intersect_merge.py#L805

Reply:
Do these distances come from a `match_annotations_two_sources()` call?

Reply:
Yes, in the above line, it gets from self._match_segments which calls match annotations two sources

Reply:
Please refactor the code, it should not work this way.